### ISSUE TITLE
Memory optimization for CaptureFromCamera with offline rendering mode

### ISSUE THREAD

**Issue by PhannGor** [RenderHeads/UnityPlugin-AVProMovieCapture#74]
We're using CaptureFromCamera in offline mode together with UnityAudioCapture (From AudioRenderer) to capture high-quality videos on iOS devices. We have pretty intense scenes that consume a lot of RAM by themselves so we're looking for any possible way to decrease memory consumption by capturing process.

We don't mind if rendering will take a longer time than usual, our main goal is to make rendering possible on iOS devices without the out-of-memory crashes we're experiencing now.

We did a few changes in CaptureFromCamera.cs and CaptureFromAudioRenderer.cs (inspired by your EXPERIMENTAL_FRAME_PAUSE) which improved the situation a bit:

* Modified CaptureFromCamera.FinalRenderCapture() coroutine to pause until frame encoding is done.
```
private IEnumerator FinalRenderCapture()
{
	if (_stillBusy)
	{
		yield return null;
	}
	else
	{
		_stillBusy = true;

		Time.timeScale = 0.0f;
		yield return new WaitUntil(CanContinue);
		Time.timeScale = 1.0f;

		yield return _waitForEndOfFrame;
		yield return Capture();
		_stillBusy = false;
	}
}
```

* Modified CaptureFromAudioRenderer.Update() correspondingly to not encode audio while paused:
```
void Update()
{
	if (_isRendererRecording && _capture != null && _capture.IsCapturing() && Time.timeScale > 0)
	{
		int sampleFrameCount = AudioRenderer.GetSampleCountForCaptureFrame();
		int sampleCount = sampleFrameCount * _unityAudioChannelCount;
		// TODO: reuse NativeArray for less GC (but not super important in offline mode)
		NativeArray<float> audioSamples = new NativeArray<float>(sampleCount, Allocator.TempJob);
		if (AudioRenderer.Render(audioSamples))
		{
			// TODO: use NativeArray instead of converting to array for less GC (but not super important in offline mode)
			_capture.EncodeAudio(audioSamples.ToArray());
		}
		audioSamples.Dispose();
	}
}
```

But this still not enough for our goals. For example, for iPhone XS it works fine for 1080p with both video and audio. Regarding 4K, we're able to encode 4K videos on iPhone XS when audio disabled. However,  it constantly crashes when we're trying to record 4K with audio. Audio encoding adds something like 150-200 Mb memory overhead in our case. 

So as I said before,  we're looking for any possible way to decrease memory consumption by capturing process. Either audio or video encoding, or both. And we would appreciate it if you could point us in the right direction or for any help you could provide regarding that question.

Unity version: 2019.4.10f1
AVPro Movie Capture version: 4.4.1
Operating system version: iOS 13.7 (iPhone XS)
Capture component used: CaptureFromCamera
Capture settings (resolution, frame-rate, codec): 4k, 30 FPS, H264

**Comment by AndrewRH** [comment]
Hi @PhannGor 

Thanks for these tips.

We have now added the pause check into CaptureAudioFromAudioRenderer.

I'm not sure what you changed in the coroutine besides using it and refactoring the way the yield is written, but we've made those changes too.  This is indeed a path we're considering adding in the future but currently it is still defined out while we test it more.

There is a new version of  plugin (4.4.3) which will be released today which includes these changes and a ton of other fixes for macOS and iOS.  There is a small chance that this new version will have improved memory usage, so please try it and let us know.

Unfortunately the Apple platform just seems to use a lot of memory in offline mode when using audio.  This seems to because it prioritises audio, and so it buffers all of the video frames until it has 1 second of audio.  As you can imagine this ends up being quite a lot of memory.

Try the new version and let us know.

Thanks,

**Comment by PhannGor** [comment]
Hi @AndrewRH,

Thanks for your fast reply and for the explanation of what's going on with audio encoding on iOS. Will wait for 4.4.3 then to give it a spin and will let you know about results.

Thank you!

**Comment by PhannGor** [comment]
Hi @AndrewRH,

Just updated to v4.4.3 and did few tests but unfortunately there is no noticeable difference in memory consumption. Is there anything else we could try or at least could you point us in which direction we should look to improve memory usage. We don't care so much about rendering speed, it could be pretty slow but should not consume too much memory.

Thank you!

**Comment by AndrewRH** [comment]
Oh that's a shame.   Yes as I said before, I'm not sure there is much we can do as this just seems to be what AVFoundation does.   I'm not sure if @MorrisRH has any more to add...

FYI We also released version 4.4.4 which has some minor fixes.

Thanks,

**Comment by PhannGor** [comment]
Thank you @AndrewRH. I understand the situation and had the same thoughts but it was worth asking you guys first. I guess we can close the issue then.
